### ISSUE TITLE
Refactor out cmd/syncthing

### ISSUE THREAD

**Issue by calmh** [syncthing/syncthing#4085]
Currently cmd/syncthing contains a shitload of stuff. This makes sense as is, but complicates reuse. What reuse? Possibilities:

 - libsyncthing.so which is actually loadable from C land and contains all the required functionality.
 - An actual GUI application written as above or in Go directly (using Qt bindings for example)
 - The current CLI/web application.

Ideally cmd/syncthing would be the 100 or so lines requires to parse environment variables and command line switches, maybe load the config, and then `app := syncthing.NewApp(&cfg); app.Run()` or something like that.

**Comment by Martchus** [comment]
Note only a few adjustments are required to make the `libsyncthing.so` use-case *basically* work:
* adjustments in Syncthing's repo itself: https://github.com/Martchus/syncthing/tree/libsyncthing
* C++ interface and CMake integration for building Syncthing as library (contains the commands one need to invoke to build Syncthing as library): https://github.com/Martchus/syncthingtray/tree/master/libsyncthing

(I didn't propose a PR yet because I want to see whether this actually leads to something before. Also my Go knowledge isn't the best so the code quality is likely not the best, too.)

One thing which is missing for the `libsyncthing.so` use-case is the ability to stop all ongoing background tasks (goroutines). This is required to restart the Syncthing instance running 'inside' the application. So when refactoring, it would be nice if that requirement would be considered.

**Comment by AudriusButkevicius** [comment]
We should weed out the goroutines that did not shut down nicely after a stop

**Comment by Martchus** [comment]
Those should not be hard to find. In my tests I just added a sleep() at the end and stopped the debugger. I guess all threads it shows (beside the main thread) are goroutines. Except one which is the garbage collector. At least I assume it also lives in its own thread.

Btw, not sure how to keep that beast under control. `cgo` apparently does some magic to start the gc on the first C-to-Go function call. Not sure when it stops the gc again. Maybe when all goroutines terminate? But I'm just asking out of interest. I don't think this really matters in practice as long as the gc doesn't cause crashes or unwanted behavior.

**Comment by calmh** [comment]
The go runtime never terminates threads. You can count the number of goroutines though with something like runtime.NumGoRoutines() (which should be 1 if the main goroutine is the only one still alive and that’s where you’re doing the call).

**Comment by calmh** [comment]
In addition to the already mentioned shutdown/restart issue we should also remove calls to log.Fatal and similar (these should instead be a return from the top level function with an error of some kind).

The logger also merits some though. At the very least we need to be able to not spew stuff to stdout, as stdout may not be "ours"; better would be an injectable logger so the output could be handled however by the caller.

**Comment by Martchus** [comment]
> At the very least we need to be able to not spew stuff to stdout, as stdout may not be "ours"

I'd embrace that feature, too. However, I remember that is was possible to hack around it so for me stopping would have priority.

**Comment by imsodin** [comment]
After #5805 I see the following issues that need resolving. Please edit or let me know about more or corrections:

- [ ] Events are global.
- [ ] Connections registry is global.
- [ ] Logging is global.

@Martchus If you work on adapting your C wrapping soon, please write if there's anything to correct/improve on #5805. It would be nice if we can iron out the added interface during RC.

**Comment by Martchus** [comment]
> It would be nice if we can iron out the added interface during RC.

I actually see no reason to put time pressure on this. Your recent interface changes already look like a change in the right direction. Maybe there are things missing but that can be improved later on.

To solve the logging problem I did some hack which allowed me to send all the output to a callback which can be consumed by the C++ program and then displayed in the GUI.

I had to do some general fixes regarding the launcher first so I haven't had the time to look into the library yet. I only rebased my additions on your latest changes but haven't actually adapted them yet.

**Comment by imsodin** [comment]
I didn't intend to put time pressure on you, sorry if that came across the wrong way. It was intended as a info that if you were anyway working on it and something would come up, I'd be interested in that feedback to potentially already fix it in the RC process. Anyways thanks for your contribution!

**Comment by Martchus** [comment]
I based my basic C-bindings and integration into Syncthing Tray now on the latest master. It basically works and I didn't need to modify any files of Syncthing itself anymore. I only added my own files for the C-bindings but those could also be moved to my own repo now.

See  
* https://github.com/Martchus/syncthingtray/commit/038225936eb0ec8d9515d2d0b0d3e375d219e894
* https://github.com/Martchus/syncthing/commit/51440c77d846bbca1de48d58d6fc5ad69758ca7a

I've noticed that there are still a few things in `main.go` which would be desirable to have in the library as well. Checkout `c_bindings.go` of my tiny C wrapper and you will spot the code I've just copied and pasted from `main.go`. That is of course not a good solution. I haven't copied over code for handling previous config file versions and cleaning up the config directory but that would be useful, too.

Stopping and starting again now also doesn't produce crashes anymore. It only seems that Syncthing isn't able to stop "in a timely manner":

```
[INFO]    syncthing unknown-dev "Fermium Flea" (go1.12.6 linux-amd64) unknown@unknown 1970-01-01 00:00:00 UTC
[INFO]    My ID: H6NY7TD-3YFH6OK-G6X3YF7-HXZ7ORK-NG4THAL-TGTGZH6-YDO2XRM-HCZZZQ2
[INFO]    Single thread SHA256 performance is 402 MB/s using minio/sha256-simd (342 MB/s using crypto/sha256).
[INFO]    Hashing performance is 335.17 MB/s
[INFO]    Starting deadlock detector with 20m0s timeout
[INFO]    No stored folder metadata for "default": recalculating
[INFO]    Ready to synchronize "Default Folder" (default) (sendreceive)
[INFO]    Overall send rate is unlimited, receive rate is unlimited
[INFO]    Using discovery server https://discovery.syncthing.net/v2/?noannounce&id=LYXKCHX-VI3NYZR-ALCJBHF-WMZYSPK-QG6QJA3-MPFYMSO-U56GTUK-NA2MIAW
[INFO]    Relay listener (dynamic+https://relays.syncthing.net/endpoint) starting
[INFO]    Using discovery server https://discovery-v4.syncthing.net/v2/?nolookup&id=LYXKCHX-VI3NYZR-ALCJBHF-WMZYSPK-QG6QJA3-MPFYMSO-U56GTUK-NA2MIAW
[INFO]    TCP listener ([::]:46099) starting
[INFO]    Using discovery server https://discovery-v6.syncthing.net/v2/?nolookup&id=LYXKCHX-VI3NYZR-ALCJBHF-WMZYSPK-QG6QJA3-MPFYMSO-U56GTUK-NA2MIAW
[INFO]    Completed initial scan of sendreceive folder "Default Folder" (default)
[INFO]    Loading HTTPS certificate: open /hdd/build/qtcreator/build-syncthingtray-Desktop_Clang-Debug/syncthingtray/tray/libst/https-cert.pem: no such file or directory
[INFO]    Creating new HTTPS certificate
[INFO]    GUI and API listening on 127.0.0.1:41701
[INFO]    Access the GUI via the following URL: http://127.0.0.1:41701/
[INFO]    My name is "linux-9lzf"
[INFO]    Sent usage report (version 3)
[INFO]    Detected 1 NAT service
[INFO]    Joined relay relay://147.175.187.143:993
[INFO]    TCP listener ([::]:46099) shutting down
[INFO]    c.S.listenerSupervisor: Service dynamic+https://relays.syncthing.net/endpoint failed to terminate in a timely manner
[INFO]    connections.Service: Service c.S.listenerSupervisor failed to terminate in a timely manner
[INFO]    Exiting

Syncthing exited with exit code 0
```

By the way, that log was copied from the text area in the launcher settings of Syncthing Tray. So the logging callback actually works fine and it isn't even much code: https://github.com/Martchus/syncthing/commit/51440c77d846bbca1de48d58d6fc5ad69758ca7a#diff-68f5a820655f0f2818ba4930885a8bf7R36

**Comment by imsodin** [comment]
Very nice :)


> I've noticed that there are still a few things in `main.go` which would be desirable to have in the library as well.

Totally see that. The motivation for not including it was, that `App` should be filesystem agnostic. That's of course not yet fully true (many of it's dependencies aren't FS agnostic), but the general intention. So what I propose to do is extract this code from cmd/syncthing and put it as utility functions into the corresponding `lib/...` package and make those available in `lib/syncthing` (or put them there directly).

> Stopping and starting again now also doesn't produce crashes anymore. It only seems that Syncthing isn't able to stop "in a timely manner":

That's being addressed in #5863.

**Comment by Martchus** [comment]
Another thing: I've just tried to consume Syncthing as a Go module in order to move the additional files the my C-bindings to a different repo. That basically works but `lib/auto/gui.files.go` is missing. I've tried to adjust my CMake step which invokes `script/genassets.go` for generating the file to put it into the directory where I suppose Go would look for it. However, that didn't work because this directory is read-only. Mitigating the permission problem seemed very hacky so I didn't continue. Not sure how to solve this. Of course the c-bindings could also just live in the official repository at the point they are "nice enough".

Likely C-bindings and using Syncthing as a Go module is a bit out-of-scope of this issue but I wanted to mention it. It is also not really required for what I intent to do (the current approach of using a Git submodule works just fine).

**Comment by imsodin** [comment]
This might not be C specific. From the `go generate` docs:

>  Go generate is never run automatically by go build, go get, go test, and so on. It must be run explicitly. 

As I understand it if you want to use Syncthing as a library through the usual go way (just import it), these files will never be generated. So we need to commit those files to VCS to make it work. In my opinion that's ugly, but from a little bit of reading around this seems to be by design in golang. What are your thoughts on this @AudriusButkevicius @calmh ?

**Comment by AudriusButkevicius** [comment]
I am sure we're not the only ones that have a library with generated files, we should check what others do. I suspect they check in generated files.

**Comment by calmh** [comment]
Usually one does. I don't necessarily want to do that in our case though because these are large and unwieldy. I think we should refactor so that a set of assets, or an asset directory, or both, is passed into the constructor of the API package. It then becomes the callers responsibility. In our cmd/syncthing we can do what we do now. If there are other users, they can do what works in their use case.

**Comment by Martchus** [comment]
Checking those files in doesn't seem like a nice solution indeed. The solution proposed by @calmh seems good. I needed to add a build step for the asset generation on my side anyways (which is trivial in CMake) so this wouldn't increase the effort much.
